src/Optimization/hiopNlpFormulation.cpp

```diff
@@ -480,6 +482,38 @@ bool hiopNlpFormulation::finalizeInitialization()
   return bret;
 }
 
+bool hiopNlpFormulation::apply_scaling(hiopVector& c, hiopVector& d, hiopVector& gradf, 
+                                       hiopMatrix& Jac_c, hiopMatrix& Jac_d)
+{
+  //check if we need to do scaling
+  if("none" == options->GetString("scaling_type")) {
+    return false;
+  }
+  
+  const double max_grad = options->GetNumeric("scaling_max_grad");
+
+  double obj_scale = max_grad/gradf.infnorm();
+  if(obj_scale>1.)
+  {
+    obj_scale=1.;
+    return false;
+  }
+
+  nlp_scaling = new hiopNLPObjGradScaling(max_grad, c, d, gradf, Jac_c, Jac_d);
```

Comment:
`apply_scaling` is called each time the starting procedure is called, which can be called multiple times when restarting the nlp. I've looked at the code and I think `nlp_scaling` will be added each time the nlp is restarted... looks like you may have forgotten to test `nlp_scaling` for null. 

this raises the questions whether the scaling should be recomputed when the nlp is restarted ...

Reply:
See [line 196](https://github.com/LLNL/hiop/blob/5e6b92008fbc998f200dbfd9ceb341f0494abb47/src/Optimization/hiopNlpFormulation.cpp#L196)
I think it release the memory in object nlp_transformations, and all the transformations, e.g., fixedVarsRelaxer and one I am working on for issue #179, are added each time when the nlp is restarted.

Note that [line 196](https://github.com/LLNL/hiop/blob/5e6b92008fbc998f200dbfd9ceb341f0494abb47/src/Optimization/hiopNlpFormulation.cpp#L196) will make pointer nlp_scaling as a dangling pointer, and then it is reset at  [line 502](https://github.com/LLNL/hiop/blob/5e6b92008fbc998f200dbfd9ceb341f0494abb47/src/Optimization/hiopNlpFormulation.cpp#L502)

Reply:
I think `nlp_transformations.clear();` on line [196](https://github.com/LLNL/hiop/blob/dev-scaling/src/Optimization/hiopNlpFormulation.cpp#L196) will always delete
Same question apply to other transformations, e.g., relax bounds for fixed variables, and relax var/con bounds before solving the problem (in the new branch I am working on).
Currently I can see you reset object `nlp_transformations` and add all the transformations back to it when nlp is restarted.